flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGFailoverTest.java

```diff
@@ -0,0 +1,763 @@
+		final JobGraph jobGraph = createBatchJobGraph();
+		final ExecutionGraph eg = createExecutionGraph(jobGraph);
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+		final ExecutionVertex ev22 = vertexIterator.next();
+		final ExecutionVertex ev31 = vertexIterator.next();
+		final ExecutionVertex ev32 = vertexIterator.next();
+		final ExecutionVertex ev41 = vertexIterator.next();
+		final ExecutionVertex ev42 = vertexIterator.next();
+
+		// trigger ev21, ev22 scheduling on data from ev31 consumable
+		testMainThreadUtil.execute(() -> {
+			eg.scheduleOrUpdateConsumers(new ResultPartitionID(
+				ev11.getProducedPartitions().keySet().iterator().next(),
+				ev11.getCurrentExecutionAttempt().getAttemptId()));
+			eg.scheduleOrUpdateConsumers(new ResultPartitionID(
+				ev12.getProducedPartitions().keySet().iterator().next(),
+				ev12.getCurrentExecutionAttempt().getAttemptId()));
+
+			// finish upstream regions to trigger scheduling of downstream regions
+			ev11.getCurrentExecutionAttempt().markFinished();
+			ev12.getCurrentExecutionAttempt().markFinished();
+			ev21.getCurrentExecutionAttempt().markFinished();
+			ev22.getCurrentExecutionAttempt().markFinished();
+
+			// trigger ev41 scheduling on data from ev31 consumable
+			eg.scheduleOrUpdateConsumers(new ResultPartitionID(
+				ev31.getProducedPartitions().keySet().iterator().next(),
+				ev31.getCurrentExecutionAttempt().getAttemptId()));
+		});
+
+		// verify initial state
+		assertVertexInState(ExecutionState.FINISHED, ev11);
+		assertVertexInState(ExecutionState.FINISHED, ev12);
+		assertVertexInState(ExecutionState.FINISHED, ev21);
+		assertVertexInState(ExecutionState.FINISHED, ev22);
+		assertVertexInState(ExecutionState.DEPLOYING, ev31);
+		assertVertexInState(ExecutionState.DEPLOYING, ev32);
+		assertVertexInState(ExecutionState.DEPLOYING, ev41);
+		assertVertexInState(ExecutionState.CREATED, ev42);
+
+		// verify initial attempt number
+		assertEquals(0, ev11.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev12.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev21.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev22.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev31.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev32.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev41.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev42.getCurrentExecutionAttempt().getAttemptNumber());
+
+		// trigger task failure of ev31 on consuming data from ev21
+		// vertices { ev11, ev21, ev31, ev32, ev41, ev42} should be affected
+		testMainThreadUtil.execute(() -> ev31.getCurrentExecutionAttempt().fail(new PartitionConnectionException(
```

Comment:
We are mostly testing the logic of `RestartPipelinedRegionStrategy`. A test like this is not justified if we want to test that `cause` is correctly propagated to `getTasksNeedingRestart()`.

Reply:
Will simplify the test to test the cause propagation only.